### ISSUE TITLE
save_resume appears to be broken

### ISSUE THREAD

**Issue by denik** [JohnLangford/vowpal_wabbit#1355]
If I train a model with --save_resume, I get one prediction:

```
$ rm -f tmp.model
$ head -n 1 test/train-sets/rcv1_micro.dat | vw --quiet --save_resume -f tmp.model
$ head -n 2 test/train-sets/rcv1_micro.dat | tail -n 1 | vw -i tmp.model -t  -p /dev/stdout --quiet
-0.098802
```

If I train a model without --save_resume, I get another prediction:

```
$ rm -f tmp.model
$ head -n 1 test/train-sets/rcv1_micro.dat | vw --quiet -f tmp.model
$ head -n 2 test/train-sets/rcv1_micro.dat | tail -n 1 | vw -i tmp.model -t -p /dev/stdout --quiet
-0.112613
```

**Comment by arielf** [comment]
@denik 
Thanks for the catch!
Reproducible 100%.

**Comment by JohnLangford** [comment]
There seems to be a real bug but this is not the relevant test.  The
relevant test compares the first sequence with:

$ head -n 2 test/train-sets/rcv1_micro.dat | vw -p /dev/stdout
--quiet

0
-0.112613

I'll try to trace things down when I get a moment.  It appears the
inconsistency is due to -t.

-John

**Comment by JohnLangford** [comment]
This is fixed in #1361 and will be merged shortly.

**Comment by JohnLangford** [comment]
The bug was a contamination of the model state with learning state when reading a save_resume model into a test only model.

**Comment by denik** [comment]
I believe there are more issues with --save_resume, see the test code in: https://github.com/JohnLangford/vowpal_wabbit/pull/1362

When using loss function / learning_rate / l1 / bootstrap  the models that are produced with and without save_resume are different: https://gist.github.com/denik/06a0d17d679dce0496e99f441bb1ceef

The loss function & learning_rate cases can be fixed by specifying training options on the follow up retrainings. However, this "fix" is not universal: in some cases it produces "duplicate option" warnings and in some cases (bootstrap, l1) the results are still different: https://gist.github.com/denik/c1de6d307432c9350f6680e1fc76f8e6

**Comment by JohnLangford** [comment]
Critical-for-testing parameters are stored in the model itself while others
(loss function/learning rate/l1/bootstrap are not.    If you respecify such
an option you get duplicate option warnings.

The --bootstrap issue sounds like a real bug.  I'm unsure about --l1, but
that may be as well.

-John

**Comment by denik** [comment]
OK. Somehow I assumed that 'vw --save_resume' would store all options, so that previous learning can continue with only "vw -i model".

I've added a few more tests to #1362. Besides --l1 and --bootstrap, --ftrl and --pistol also fail (with --repeat_args).
